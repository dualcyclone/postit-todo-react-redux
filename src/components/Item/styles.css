li {
    padding: 5px;
    margin: 1em;
    float: left;
    list-style: none;
}

li div {
    text-decoration: none;
    font-family: 'Nothing You Could Do', cursive;
    color: #000;
    display: block;
    height: 10em;
    width: 10em;
    padding: 1em;
    box-shadow: 5px 5px 7px rgba(33,33,33,.7);
    transition: -webkit-transform .15s linear;
    transform: rotate(-6deg);
    overflow: hidden;
}

li.complete div p {
    text-decoration: line-through;
}

li.complete div p:before {
    position: absolute;
    content: '';
    background: #000;
    display: block;
    width: 100%;
    height: 1px;
    transform: rotate(-50deg);
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;
    margin: auto;
}

li.complete div p:after {
    position: absolute;
    content: '';
    background: #000;
    display: block;
    width: 100%;
    height: 1px;
    transform: rotate(50deg);
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;
    margin: auto;
}

li:nth-child(even) div {
    transform: rotate(4deg);
    position: relative;
    top: 5px;
}

li:nth-child(3n) div {
    transform: rotate(-3deg);
    position: relative;
    top: -5px;
}

li:nth-child(4n) div {
    transform: rotate(2deg);
    position: relative;
    top: -5px;
}

li:nth-child(5n) div {
    transform: rotate(5deg);
    position: relative;
    top: -10px;
}

li div:hover, li div:focus{
    box-shadow: 10px 10px 7px rgba(0,0,0,.7);
    transform: scale(1.25);
    position: relative;
    z-index: 5;
}

li div .icon {
    display: block;
    position: absolute;
    visibility: hidden;
}

li div:hover .icon {
    visibility: visible;
}

li div .icon {
    color: #000;
}

li div .icono-trash {
    bottom: 2px;
    right: 2px;
}

li div .icono-check, li div .icono-cross {
    bottom: 2px;
    left: 2px;
}

li.complete div .icono-check {
    display: none;
}

li.todo div .icono-cross {
    display: none;
}
