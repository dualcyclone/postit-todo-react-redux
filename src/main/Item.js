import randomColor from 'randomcolor';

export default class Item {
  constructor(id, content) {
    return {
      id: id,
      content: content,
      completed: false,
      color: randomColor({ luminosity: 'light' })
    };
  }
};
