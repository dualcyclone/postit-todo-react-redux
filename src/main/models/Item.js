import randomColor from 'randomcolor'

export class Item {
  static generateKey() {
    return btoa(Date.now())
  }

  constructor(content) {
     return {
      id: Item.generateKey(),
      content: content,
      completed: false,
      color: randomColor({ luminosity: 'light' })
    }
  }
}

export default Item
