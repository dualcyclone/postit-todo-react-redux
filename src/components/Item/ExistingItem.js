import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { completeItem, removeItem } from '../../main/store/actions'
import Item from './Item'
import './styles.css'

export const ExistingItem = ({ item, onDoneClick, onRemoveClick }) => (
  <Item
    className={ item.completed ? 'complete' : 'todo' }
    color={ item.color }>
      <p>{ item.content }</p>
      <a onClick={onDoneClick} className="icon icono-cross">&nbsp</a>
      <a onClick={onDoneClick} className="icon icono-check">&nbsp</a>
      <a onClick={onRemoveClick} className="icon icono-trash">&nbsp</a>
  </Item>
)

ExistingItem.propTypes = {
  item: PropTypes.object.isRequired,
  onDoneClick: PropTypes.func.isRequired,
  onRemoveClick: PropTypes.func.isRequired
}

const mapDispatchToProps = (dispatch, ownProps) => ({
  onDoneClick: () => dispatch(completeItem(ownProps.item)),
  onRemoveClick: () => dispatch(removeItem(ownProps.item))
})

export default connect(
  null,
  mapDispatchToProps
)(ExistingItem)
