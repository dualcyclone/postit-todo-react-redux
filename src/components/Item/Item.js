import React from 'react';
import PropTypes from 'prop-types';
import './styles.css';

const Item = ({ ElementType = 'li', color, children, className }) => (
  <ElementType className={ 'todoItem ' + className }>
    <div style={{ backgroundColor: color }}>
      { children }
    </div>
  </ElementType>
);

Item.propTypes = {
  className: PropTypes.string.isRequired, 
  children: PropTypes.array.isRequired
};

export default Item;
