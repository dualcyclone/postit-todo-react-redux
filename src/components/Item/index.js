import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { completeItem, removeItem } from '../../main/actions';
import './styles.css';

const Item = ({ item, children, onDoneClick, onRemoveClick }) => (
  <li className={ item.completed ? 'complete' : 'todo' }>
    <div style={{ backgroundColor: item.color }}>
      <p>{ children }</p>
      <a onClick={onDoneClick} className="icon icono-cross">&nbsp;</a>
      <a onClick={onDoneClick} className="icon icono-check">&nbsp;</a>
      <a onClick={onRemoveClick} className="icon icono-trash">&nbsp;</a>
    </div>
  </li>
);

Item.propTypes = {
  item: PropTypes.object.isRequired,
  children: PropTypes.string.isRequired,
  onDoneClick: PropTypes.func.isRequired,
  onRemoveClick: PropTypes.func.isRequired
};

const mapDispatchToProps = (dispatch, ownProps) => ({
  onDoneClick: () => dispatch(completeItem(ownProps.item)),
  onRemoveClick: () => dispatch(removeItem(ownProps.item))
});

export default connect(
  null,
  mapDispatchToProps
)(Item);
