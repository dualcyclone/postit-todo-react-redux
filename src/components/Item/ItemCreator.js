import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { addItem } from '../../main/actions';
import Item from './Item';
import './styles.css';

export class ItemCreator extends Component {
  constructor(props) {
    super(props);
    this.state = {
      inputValue: ''
    };
  }

  render() {
    return (
      <Item ElementType='div' className='itemCreator'>
        <p>Add another item...</p>
        <textarea
          ref="inputField"
          className={'itemCreator-input'}
          type="text"
          rows="8"
          placeholder="What do you need to do?"
          onKeyPress={(e) => {
            if (e.key === 'Enter') {
              e.preventDefault();
              this.props.onAdd(this.state.inputValue);
              this.setState({
                inputValue: ''
              })
            }
          }}
          value={this.state.inputValue}
          onChange={evt => this.updateInputValue(evt)}
        />
        <a onClick={() => {
            this.state.inputValue && this.props.onAdd(this.state.inputValue);
            this.setState({
              inputValue: ''
            })
          }} className='icon icono-plusCircle'>&nbsp;</a>
      </Item>
    );
  }

  updateInputValue(evt) {
    this.setState({
      inputValue: evt.target.value
    });
  }
}

ItemCreator.propTypes = {
  onAdd: PropTypes.func.isRequired
};

const mapDispatchToProps = (dispatch, ownProps) => ({
  onAdd: (newItem) => dispatch(addItem(newItem))
});

export default connect(
  null,
  mapDispatchToProps
)(ItemCreator);
