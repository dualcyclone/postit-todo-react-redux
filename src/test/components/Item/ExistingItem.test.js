import React from 'react'
import { expect } from 'chai'
import sinon from 'sinon'
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'
import { ExistingItem } from '../../../../src/components/Item/ExistingItem'
import Item from '../../../../src/components/Item/Item'

const jestExpect = global.expect

configure({adapter: new Adapter()})

const defaultProps = {
  item: {
    completed: false,
    color: '#456',
    content: 'TEST TODO'
  },
  onDoneClick: sinon.stub(),
  onRemoveClick: sinon.stub()
}

describe('ExistingItem', () => {
  it('Should render without failing', () => {
    const renderedItem = shallow(<ExistingItem {...defaultProps} />)

    jestExpect(toJson(renderedItem)).toMatchSnapshot()
  })

  it('Should render with the class of "todo"', () => {
    const renderedItem = shallow(
      <ExistingItem {...defaultProps} />
    )

    expect(renderedItem.is('.todo')).to.equal(true)
    expect(renderedItem.is('.complete')).to.equal(false)
  })

  it('Should render with the class of "complete" if the item is marked as complete', () => {
    const completedItem = Object.assign({}, defaultProps.item, {
      completed: true
    })
    const propsWithCompleteItem = Object.assign({}, defaultProps, {
      item: completedItem
    })
    const renderedItem = shallow(
      <ExistingItem {...propsWithCompleteItem} />
    )

    expect(renderedItem.is('.todo')).to.equal(false)
    expect(renderedItem.is('.complete')).to.equal(true)
  })

  it('Should implement a single Item', () => {
    const renderedItem = shallow(
      <ExistingItem {...defaultProps} />
    )

    expect(renderedItem.find(Item).length).to.equal(1)
  })

  it('Should provide a colour where one is provided', () => {
    const renderedItem = shallow(
      <ExistingItem {...defaultProps} />
    )

    expect(renderedItem.find(Item).props().color).to.equal(defaultProps.item.color)
  })

  describe('Should contain the following children', () => {
    it('The first implementing child of the component should contain the items content', () => {
      const renderedItem = shallow(
        <ExistingItem {...defaultProps} />
      )

      expect(renderedItem.find(Item).find('p').props().children).to.equal(defaultProps.item.content)
    })

    it('The remaining children should be buttons that implement onDoneClick and onRemoveClick properties (not completed)', () => {
      const renderedItem = shallow(
        <ExistingItem {...defaultProps} />
      )

      expect(renderedItem.find(Item).find('button').get(0).props.className).to.equal('icon icono-check')
      expect(renderedItem.find(Item).find('button').get(0).props.onClick).to.equal(defaultProps.onDoneClick)

      expect(renderedItem.find(Item).find('button').get(1).props.className).to.equal('icon icono-trash')
      expect(renderedItem.find(Item).find('button').get(1).props.onClick).to.equal(defaultProps.onRemoveClick)
    })

    it('The remaining children should be buttons that implement onDoneClick and onRemoveClick properties (completed)', () => {
      const renderedItem = shallow(
        <ExistingItem {...defaultProps} item={{ ...defaultProps.item, completed: true }} />
      )

      expect(renderedItem.find(Item).find('button').get(0).props.className).to.equal('icon icono-cross')
      expect(renderedItem.find(Item).find('button').get(0).props.onClick).to.equal(defaultProps.onDoneClick)

      expect(renderedItem.find(Item).find('button').get(1).props.className).to.equal('icon icono-trash')
      expect(renderedItem.find(Item).find('button').get(1).props.onClick).to.equal(defaultProps.onRemoveClick)
    })
  })
})
