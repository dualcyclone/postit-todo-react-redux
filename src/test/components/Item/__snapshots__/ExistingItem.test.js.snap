exports[`ExistingItem Should render without failing 1`] = `
<Component
  className="todo"
  color="#456">
  <p>
    TEST TODO
  </p>
  <a
    className="icon icono-cross"
    onClick={[Function]}>
     
  </a>
  <a
    className="icon icono-check"
    onClick={[Function]}>
     
  </a>
  <a
    className="icon icono-trash"
    onClick={[Function]}>
     
  </a>
</Component>
`;
